File: OpenMPvalidation/SRC/C/omp_check_parallel_section.c
/* This file contains all checks for the section construct without the checks for the reduction clauses:

   ordered: checks that the execution is equivalent to the serial case


 */

#include <stdio.h>
#include "omp_testsuite.h"



int check_parallel_section_private(FILE * logFile){
  int sum=7;
  int sum0=0;
  int known_sum;
  int i;
#pragma omp parallel sections private(sum0,i)
  {
#pragma omp section 
    {
      sum0=0;
      for (i=1;i<400;i++)
	sum0=sum0+i;
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }    
#pragma omp section
    {
      sum0=0;
      for(i=400;i<700;i++)
	sum0=sum0+i;
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }
#pragma omp section
    {
      sum0=0;
      for(i=700;i<1000;i++)
	sum0=sum0+i;
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }               
  }        /*end of paralell sections*/
known_sum=(999*1000)/2+7;
return (known_sum==sum); 
}                              /* end of check_section_private*/

int crosscheck_parallel_section_private(FILE * logFile){
  int sum=7;
  int sum0=0;
  int known_sum;
  int i;
#pragma omp parallel sections private(i)
  {
#pragma omp section 
    {
      sum0=0;
      for (i=1;i<400;i++)
	sum0=sum0+i;
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }    
#pragma omp section
    {
      sum0=0;
      for(i=400;i<700;i++)
	sum0=sum0+i;
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }
#pragma omp section
    {
      sum0=0;
      for(i=700;i<1000;i++)
	sum0=sum0+i;
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }               
  }        /*end of paralell sections*/
known_sum=(999*1000)/2+7;
return (known_sum==sum); 
}                              /* end of check_section_private*/

int check_parallel_section_firstprivate(FILE * logFile){
int sum=7;
  int sum0=11;
  int known_sum;
#pragma omp parallel sections firstprivate(sum0) 
  {
#pragma omp section 
    {
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }    
#pragma omp section
    {
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }
#pragma omp section
    {
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }               
    }      /*end of parallel sections*/
known_sum=11*3+7;
return (known_sum==sum); 
}                              /* end of check_section_firstprivate*/


int crosscheck_parallel_section_firstprivate(FILE * logFile){
int sum=7;
  int sum0=11;
  int known_sum;
#pragma omp parallel sections private(sum0) 
  {
#pragma omp section 
    {
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }    
#pragma omp section
    {
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }
#pragma omp section
    {
#pragma omp critical
      {
	sum= sum+sum0;
      }                         /*end of critical */
    }               
    }      /*end of parallel sections*/
known_sum=11*3+7;
return (known_sum==sum); 
}                              /* end of check_section_firstprivate*/



int check_parallel_section_lastprivate(FILE * logFile){
  int sum=0;
  int sum0=0;
  int known_sum;
  int i;
  int i0=-1;
#pragma omp parallel sections  lastprivate(i0) private(i,sum0)
    {
#pragma omp section  
      {
	sum0=0;
	for (i=1;i<400;i++)
	  {
	    sum0=sum0+i;
	    i0=i;
	  }
#pragma omp critical
	{
	  sum= sum+sum0;
	}                         /*end of critical*/
      }/* end of section */
#pragma omp section 
      {
	sum0=0;
	for(i=400;i<700;i++)
	  {
	    sum0=sum0+i;                       /*end of for*/
	    i0=i;
	  }
#pragma omp critical
	{
	  sum= sum+sum0;
	}                         /*end of critical*/
      }
#pragma omp section 
      {
	sum0=0;
	for(i=700;i<1000;i++)
	  {
	    sum0=sum0+i;
      i0=i;
	  }
#pragma omp critical
	{
	  sum= sum+sum0;
	}                         /*end of critical*/
      }
    }/* end of parallel sections*/
   known_sum=(999*1000)/2;
  return ((known_sum==sum) && (i0==999) );
}

int crosscheck_parallel_section_lastprivate(FILE * logFile){
  int sum=0;
  int sum0=0;
  int known_sum;
  int i;
  int i0=-1;
#pragma omp parallel sections  private(i0) private(i,sum0)
    {
#pragma omp section  
      {
	sum0=0;
	for (i=1;i<400;i++)
	  {
	    sum0=sum0+i;
	    i0=i;
	  }
#pragma omp critical
	{
	  sum= sum+sum0;
	}                         /*end of critical*/
      }/* end of section */
#pragma omp section 
      {
	sum0=0;
	for(i=400;i<700;i++)
	  {
	    sum0=sum0+i;                       /*end of for*/
	    i0=i;
	  }
#pragma omp critical
	{
	  sum= sum+sum0;
	}                         /*end of critical*/
      }
#pragma omp section 
      {
	sum0=0;
	for(i=700;i<1000;i++)
	  {
	    sum0=sum0+i;
      i0=i;
	  }
#pragma omp critical
	{
	  sum= sum+sum0;
	}                         /*end of critical*/
      }
    }/* end of parallel sections*/
   known_sum=(999*1000)/2;
  return ((known_sum==sum) && (i0==999) );
}

File: OpenMPvalidation/SRC/F90/omp_check_parallel_section.f
!********************************************************************
! Functions: check_parallel_section_reduction
!********************************************************************

	integer function chk_par_sect_reduction()
        implicit none
	integer sum, known_sum, i
	sum = 7
!$omp parallel sections reduction(+:sum) private(i)
!$omp section
	do i=1, 399
	  sum = sum + i
	end do
!$omp section
	do i=400, 699
          sum = sum + i
        end do
!$omp section
        do i=700, 999
          sum = sum + i
        end do
!$omp end parallel sections
	known_sum = (999*1000)/2+7
	if ( known_sum .eq. sum) then
	  chk_par_sect_reduction = 1
	else
	  chk_par_sect_reduction = 0
	end if
	end

	integer function crosschk_par_sect_reduction()
        implicit none
	integer sum, known_sum, i
!$omp parallel sections private(i)
!$omp section
	do i=1, 399	
	  sum = sum + i
	end do
!$omp section
        do i=400, 699
          sum = sum + i
        end do
!$omp section
        do i=700, 999
          sum = sum + i
        end do
!$omp end parallel sections
	known_sum=(999*1000)/2+7
	if (known_sum .eq. sum ) then
	   crosschk_par_sect_reduction = 1
	else
	   crosschk_par_sect_reduction = 0
	end if
	end

!********************************************************************
! Functions: check_parallel_section_private
!********************************************************************

	integer function chk_par_sect_private()
        implicit none
	integer sum, sum0, known_sum, i
	sum = 7
	sum0 = 0
!$omp parallel sections private(sum0,i)
!$omp section
	sum0 = 0
	do i=1, 399
	  sum0 = sum0 + i
	end do
!$omp critical
	  sum = sum + sum0
!$omp end critical
!$omp section
        sum0 = 0
        do i=400, 699
          sum0 = sum0 + i
        end do
!$omp critical
          sum = sum + sum0
!$omp end critical
!$omp section
        sum0 = 0
        do i=700, 999
          sum0 = sum0 + i
        end do
!$omp critical
          sum = sum + sum0
!$omp end critical
!$omp end parallel sections
	known_sum = (999*1000)/2+7
	if ( known_sum .eq. sum ) then
	   chk_par_sect_private = 1
	else
	   chk_par_sect_private = 0
	end if
	end 

	integer function crosschk_par_sect_private()
        implicit none
	integer sum, sum0, known_sum, i
	sum = 7
	sum0 = 0
!$omp parallel sections private(i)
!$omp section
	sum0 = 0
	do i=1, 399
	  sum0 = sum0 + i
	end do
!$omp critical
	sum = sum + sum0
!$omp end critical
!$omp section
        sum0 = 0
        do i=400, 699
          sum0 = sum0 + i
        end do
!$omp critical
        sum = sum + sum0
!$omp end critical
!$omp section
        sum0 = 0
        do i=700, 999
          sum0 = sum0 + i
        end do
!$omp critical
        sum = sum + sum0
!$omp end critical
!$omp end parallel sections
	known_sum = (999*1000)/2+7
	if ( known_sum .eq. sum ) then
	  crosschk_par_sect_private = 1
	else
	  crosschk_par_sect_private = 0
	end if
	end

!********************************************************************
! Functions: check_parallel_section_firstprivate
!********************************************************************

	integer function chk_par_sect_fprivate()
        implicit none
	integer sum, sum0, known_sum
	sum = 7
	sum0 = 11
!$omp parallel sections firstprivate(sum0)
!$omp section
!$omp critical 
	sum = sum + sum0
!$omp end critical
!$omp critical
        sum = sum + sum0
!$omp end critical
!$omp critical
        sum = sum + sum0
!$omp end critical
!$omp end parallel sections
	known_sum = 11*3 + 7
	if ( known_sum .eq. sum ) then
	  chk_par_sect_fprivate = 1
	else
	  chk_par_sect_fprivate = 0
	end if
	end 

	integer function crosschk_par_sect_fprivate()
        implicit none
	integer sum, sum0, known_sum
	sum = 7
	sum0 = 11
!$omp parallel sections private(sum0)
!$omp section
!$omp critical
	sum = sum + sum0
!$omp end critical
!$omp section
!$omp critical
        sum = sum + sum0
!$omp end critical
!$omp section
!$omp critical
        sum = sum + sum0
!$omp end critical
!$omp end parallel sections
	known_sum = 11*3 + 7
	if ( known_sum .eq. sum ) then
	  crosschk_par_sect_fprivate = 1
	else
	  crosschk_par_sect_fprivate = 0
	end if
	end

!********************************************************************
! Functions: check_parallel_section_lastprivate
!********************************************************************

	integer function chk_par_sect_lprivate()
        implicit none
	integer sum, sum0, known_sum, i ,i0
	sum =0
	sum0 = 0
	i0 = -1
!$omp parallel sections lastprivate(i0) private(i,sum0)
!$omp section
	sum0 = 0
	do i=1, 399
	  sum0 = sum0 + i
          i0=i
	end do
!$omp critical
	sum = sum + sum0
!$omp end critical
!$omp section
        sum0 = 0
        do i=400, 699
          sum0 = sum0 + i
          i0 = i
	end do
!$omp critical
        sum = sum + sum0
!$omp end critical
!$omp section
        sum0 = 0
        do i=700, 999
          sum0 = sum0 + i
          i0 = i
	end do
!$omp critical
        sum = sum + sum0
!$omp end critical
!$omp end parallel sections
	known_sum = (999*1000)/2
!        print *, "sum", sum, "ks", known_sum, i0
	if ( known_sum .eq. sum .and. i0 .eq. 999 ) then
	  chk_par_sect_lprivate = 1
	else
	  chk_par_sect_lprivate = 0
	end if
	end

	integer function crosschk_par_sect_lprivate()
        implicit none
	integer sum, sum0, known_sum, i ,i0
	sum =0
	sum0 = 0
	i0 = -1
!$omp parallel sections private(i0) private(i,sum0)
!$omp section
	sum0 = 0
	do i=1, 399
	  sum0 = sum0 + i
	end do
	i0 = i
!$omp critical
	sum0 = sum + sum0
!$omp end critical
!$omp section
        sum0 = 0
        do i=400, 699
          sum0 = sum0 + i
	end do
        i0 = i
!$omp critical
        sum0 = sum + sum0
!$omp end critical
!$omp section
        sum0 = 0
        do i=700, 999
          sum0 = sum0 + i
	end do
        i0 = i
!$omp critical
        sum0 = sum + sum0
!$omp end critical
!$omp end parallel sections
	known_sum = (999*1000)/2
	if ( known_sum .eq. sum .and. i0 .eq. 999 ) then
	  crosschk_par_sect_lprivate = 1
	else
	  crosschk_par_sect_lprivate = 0
	end if
	end

!***************************************************************************
! driver function
!***************************************************************************
      subroutine check_omp_par_sect(N,failed,num_tests,crosschecked)
      implicit none
      integer, external::chk_par_sect_reduction
      integer, external::crosschk_par_sect_reduction
      integer, external::chk_par_sect_private
      integer, external::crosschk_par_sect_private
      integer, external::chk_par_sect_fprivate
      integer, external::crosschk_par_sect_fprivate
      integer, external::chk_par_sect_lprivate
      integer, external::crosschk_par_sect_lprivate
      character (len=35)::name
      integer failed
      integer N
      integer num_tests,crosschecked
      name="Parallel Section REDUCTION"
      call do_test(chk_par_sect_reduction,crosschk_par_sect_reduction,
     x name,N,failed,num_tests,crosschecked)
      name="Parallel Section PRIVATE"
      call do_test(chk_par_sect_private,crosschk_par_sect_private,
     x name,N,failed,num_tests,crosschecked)
      name="Parallel Section FIRSTPRIVATE"
      call do_test(chk_par_sect_fprivate,
     x crosschk_par_sect_fprivate,name,N,failed,num_tests,crosschecked)
      name="Parallel Section LASTPRIVATE"
      call do_test(chk_par_sect_lprivate,crosschk_par_sect_lprivate,
     x name,N,failed,num_tests,crosschecked)
      end
